starfish/codebook.py

```diff
@@ -356,13 +357,101 @@ def to_json(self, filename: str) -> None:
         with open(filename, 'w') as f:
             json.dump(code_array, f)
 
-    def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
+    @staticmethod
+    def _normalize_features(array: Union["Codebook", IntensityTable], norm) \
+            -> Tuple[Union["Codebook", IntensityTable], np.ndarray]:
+        """unit normalize each feature of array
+
+        Parameters
+        ----------
+        array : Union[Codebook, IntensityTable]
+            codebook or intensity table containing (ch, r) features to normalize
+        norm : int
+            the norm to apply to each feature
+
+        See Also
+        --------
+        The available norms for this function can be found at the following link:
+        https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.linalg.norm.html
+
+        Returns
+        -------
+        Union[IntensityTable, Codebook] :
+            IntensityTable or Codebook containing normalized features
+        np.ndarray :
+            A 1 dimensional numpy array containing the feature norms
+
+        """
+        feature_traces = array.stack(traces=(Indices.CH.value, Indices.HYB.value))
+        norm = np.linalg.norm(feature_traces.values, ord=norm, axis=1)
+        array = array / norm[:, None, None]
+
+        # if a feature is all zero, the information should be spread across the channel
+        partitioned_intensity = 1 / np.product(array.shape[1:])
```

Comment:
this only works for an L1 norm. to generalize, you need something like ```np.linalg.norm(np.ones(N)*(1/N),ord=norm)``` where N is the length of your pixel trace